src/audio/pipeline/pipeline-stream.c

```diff
@@ -352,6 +352,20 @@ static int pipeline_comp_trigger(struct comp_dev *current,
 		 "pipeline_comp_trigger(), current->comp.id = %u, dir = %u",
 		 dev_comp_id(current), dir);
 
+#if CONFIG_IPC_MAJOR_4
+	/*
+	 * IPC4 has a SET_PIPELINE_STATE for each pipeline, so FW
+	 * should not propagate triggers on its own.
+	 * IPC3 has commands only for graph edges, so propagation is required.
+	 */
+	if (!is_single_ppl) {
+		pipe_dbg(current->pipeline,
+			 "pipeline_comp_trigger(), stop trigger propagation at comp.id = %u",
+			 dev_comp_id(current));
+		return 0;
+	}
+#endif
```

Comment:
line 403 also discuss the propagation, can we combine it ?

Reply:
This seems a rather complicated piece of machinery to me. So I think it needs good documentation. It should describe, that:
Pipeline triggering is performed in two stages:
1. from the IPC context: there we set trigger flags for all affected pipelines and submit them for scheduling. For IPC4 we already check and only do that for pipelines, explicitly requested by the host, with no propagation, as determined by https://github.com/thesofproject/sof/blob/ee80da7b00f32170715ba91302b35aad858d75fe/src/audio/pipeline/pipeline-stream.c#L197 . The pipeline order was unimportant until now, but with #8504 the IPC thread will wait when setting each flag, so the order does matter with that. And the order is as sent by the host. And the host orders them output-to-input. So for a 2-pipeline playback stream we first trigger the DAI pipeline, and only on the next tick we trigger the host pipeline. But they're added to the scheduling list in the priority order, which is from input to output, so the scheduling will be first the host and then the DAI pipeline.
2. in the pipeline scheduler context, as described above, we actually execute the trigger operation. But when doing this we trigger not only the pipeline, where we've found the flag, but all connected pipelines too, and that breaks the ordering. This PR addresses that. With it we should get:
```

- IPC: add DAI for triggering, wait
- scheduler (tick 1):
  * process trigger DAI
  * no data - no copying
- IPC: complete waiting, add host for triggering - host task scheduled before DAI
- scheduler (tick 2):
  * process trigger host
  * produce data by host
  * copy data by DAI
```

As for @RanderWang 's proposal - I agree, let's merge it with 403 into something like
```
if (!is_single_ppl && (!is_same_sched || IS_ENABLED(CONFIG_IPC_MAJOR_4))) {
```
(I assume that the `pipeline_should_report_enodata_on_trigger()` logic will still work correctly).

Reply:
@lyakh Ack on all the above, except I don't think enodata is good. It does work with current Linux SOF driver, so we can debate this, but fundamentally in IPC4, it's the host that controls when a pipeline is set RUNNING, so I just can't see how the ENODATA check can be reliably done in FW with IPC4.